### PR TITLE
[DISCO-2248] Re-use Merino Remote Settings backend module in test

### PR DESCRIPTION
**Pull request by Trinaa**
# Description
* Refactor RemoteSettingsBackend for re-use in test frameworks
    * Sort methods by execution order
    * Replace deprecated asyncio `as_completed` with TaskGroup 
    * Add exception handling
* Refactor/Add unit tests
* Replace kinto_client module with the remotesettings module in the load test framework
* Update load test Docker files to include merino modules
* Update setup_k8s
* Add cloudbuild.yaml and update setup_k8s.sh to accommodate new Docker context in distributed tests

# Issue(s)
#208 | [DISCO-2248](https://mozilla-hub.atlassian.net/browse/DISCO-2248)

[DISCO-2248]: https://mozilla-hub.atlassian.net/browse/DISCO-2248?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Trinaa** [comment]
@dlactin This change would require the inclusion of merino directory modules in the docker image. This has proven tricky because the load test framework Dockerfile is not located in the project root.
1. Do you have any feedback regarding my changes to the Dockerfile or docker-compose.yaml? Can this be done better?
2. Will this impact the load tests in CD?

**Comment by dlactin** [comment]
> @dlactin This change would require the inclusion of merino directory modules in the load test docker image. This has proven tricky because the load test framework Dockerfile is not located in the project root.
> 
>     1. Do you have any feedback regarding my changes to the Dockerfile or docker-compose.yaml? Can this be done better?
> 
>     2. Will this impact the load tests in CD?

Your dockerfile changes look good to me. 

No impact on the load tests in CD, though as discussed over slack we might be better served with a circleci job to build and publish the image to dockerhub.

**Comment by Trinaa** [comment]
Re-based and fixed some stuff I found in DISCO-2231

File: merino/providers/adm/backends/remotesettings.py [hunk 1]
```diff
@@ -34,91 +54,151 @@ def __init__(self, server: str, collection: str, bucket: str) -> None:
...
+        try:
+            async with asyncio.TaskGroup() as task_group:
+                for record in data_records:
+                    task_group.create_task(
+                        extend_suggestions_lst(
+                            url=urljoin(
+                                base=attachment_host,
+                                url=record["attachment"]["location"],
+                            ),
+                            lst=suggestions,
```
Comment (ncloudioj) [inline merino/providers/adm/backends/remotesettings.py:168]: Mutating this suggestion list across all the tasks feels a little odd to me even though Merino is single-threaded. Alternatively, you can return suggestions from each task and then await them and join the results (or use a generator).

```py
         async def extend_suggestions_lst(url: str, lst: list[KintoSuggestion]):
             return await self.get_attachment(url)

         tasks = []
         async with asyncio.TaskGroup() as task_group:
                for record in data_records:
                    tasks.append(
                        task_group.create_task(
                            extend_suggestions_lst(
                                url=urljoin(
                                    base=attachment_host,
                                    url=record["attachment"]["location"],
                             )
                          )
                      )

         for task in tasks:
            yield await task
         # Or just join all the results for simplicity
```

File: tests/load/locust_tests/locustfile.py [hunk 2]
```diff
@@ -43,7 +50,8 @@
 KINTO__COLLECTION = os.environ["KINTO__COLLECTION"]
 
 # This will be populated on each worker and
```
Comment (ncloudioj) [inline tests/load/locust_tests/locustfile.py:52]: Looks like this comment gets truncated.

**Review by ncloudioj — APPROVED** [review]
r+ wc.

Nice reuse!

File: merino/providers/adm/backends/remotesettings.py (hunk 1 above)
Comment (Trinaa) [reply merino/providers/adm/backends/remotesettings.py]: I can iterate over tasks to extend suggestions, this is clearer, ty for the suggestion!

File: tests/load/locust_tests/locustfile.py (hunk 2 above)
Comment (Trinaa) [reply tests/load/locust_tests/locustfile.py]: removed the 'and', looking through the history the addition is a mystery

**Pull request merged**